### DISCUSSION TITLE
Add external scripts or styles from plugins

### DISCUSSION THREAD

**Discussion by ahmafi** [GrapesJS/grapesjs#4594] [Q&A]
How can I add external scripts or styles from plugins? I have tried the solution on [this issue](https://github.com/artf/grapesjs/issues/66#issuecomment-293246613), but I got `getComponents` is undefined.

```js
const canvasPlugin = (editor) => {
  editor
    .getComponents()
    .add(
      '<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.1/dist/css/bootstrap.min.css">'
    );
  editor
    .getComponents()
    .add(
      '<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.1/dist/js/bootstrap.bundle.min.js"></script>'
    );
};
```

**Comment by artf** [comment]
Components are not immediately available in the plugin (eg. are loaded asynchronously from the remote storage), the safe way would be to execute that logic when the editor is ready
```js
const plugin = (editor) => {
 editor.onReady(() => {
   editor.getComponents()...
})
}
```

**Reply by ahmafi** [reply]
Thanks. It worked for styles with `<link rel="stylesheet" href="...">` but it seems I can't add `script` tags with `editor.getComponents().add()`. How can I add the script?

**Reply by ahmafi** [reply]
Is there any way to add external CSS and scripts from a plugin other than this method? Because this method is adding CSS and script as body children.

**Reply by artf** [reply]
For scripts you should enable `config.parser.optionsHtml.allowScripts` option.
You can also add your exetrnal styles/scripts in this way: https://grapesjs.com/docs/modules/Components-js.html#template-related
But you'll need then to include them manually in the exported code.

**Reply by ahmafi** [reply]
I'm building a set of components that uses external styles/scripts (bootstrap for now, but I'll use other things later), and I want to publish these components, so I can use it in many other places.
I think it would be great if I could set external styles/scripts like the second way you mentioned, but from my plugin, so I don't have to include them manually every time. What do you think? Is this a good feature to have, or there are other concerns?

**Reply by artf** [reply]
You should be able to update the config in your plugin, eg. `editor.Canvas.getConfig().styles.push('...')`

**Comment by jcenlo** [comment]
@ahmafi i had the same problem, i am trying to add bootstrap css, a solution can be in the (tailwind grapesjs plugin) check this file
[https://github.com/Ju99ernaut/grapesjs-tailwind/blob/master/src/index.js](https://github.com/Ju99ernaut/grapesjs-tailwind/blob/master/src/index.js), they solve the issue append the styles in the head of the document.
```JavaScript
  const options = {
    ...{
      i18n: {},
      // default options
      tailwindPlayCdn: 'https://cdn.tailwindcss.com',
      plugins: [],
      config: {},
      cover: `img.object-cover { filter: sepia(1) hue-rotate(190deg) opacity(.46) grayscale(.7) !important; }`,
      changeThemeText: 'Change Theme',
    }, ...opts
  };

  const appendTailwindCss = async (frame) => {
    const iframe = frame.view.getEl();

    if (!iframe) return;

    const { tailwindPlayCdn, plugins, config, cover } = options;
    const init = () => {
      iframe.contentWindow.tailwind.config = config;
    }

    const script = document.createElement('script');
    script.src = tailwindPlayCdn + (plugins.length ? `?plugins=${plugins.join()}` : '');
    script.onload = init;

    const cssStyle = document.createElement('style');
    cssStyle.innerHTML = cover;

    // checks iframe is ready before loading Tailwind CSS - issue with firefox
    const f = setInterval(() => {
      const doc = iframe.contentDocument;
      if (doc.readyState === 'complete') {
        doc.head.appendChild(script);
        doc.head.appendChild(cssStyle);
        clearInterval(f);
      }
    }, 100)
  }

  editor.Canvas.getModel()['on']('change:frames', (m, frames) => {
    frames.forEach(frame => frame.once('loaded', () => appendTailwindCss(frame)));
  });
```
